sipplauncher/Test.py

```
@@ -383,28 +382,41 @@ def post_run(self, run_id_prefix, args):
         if self.__state in [SIPpTest.State.SUCCESS, SIPpTest.State.FAIL]:
             # pre_run() has succedded.
             # Now we should attempt to cleanup as much as we can.
-            # We shouldn't propagate exception to the caller, because caller should post_run other tests as well.
+            #
+            # We should propagate exception to the caller if it's caused by internal error.
+            # This stops tests execution.
+            #
+            # We shouldn't propagate exception to the caller if it's caused by
+            # - improper test definition in test-suite
+            # - failure to cleanup the DUT
+            # User should see DIRTY state in this case and testing should continue for further tests.
+            # User could check test's logs for exception details.
             self.__set_state(SIPpTest.State.CLEANING)
             self.__print_run_state(run_id_prefix)
             start = time.time()
-            state = SIPpTest.State.CLEAN
 
-            for h in [partial(SIPpTest.__run_script, self, "after.sh", args),
-                      partial(Network.SIPpNetwork.sniffer_stop, self.network),
-                      partial(SIPpTest.__remove_temp_folder, self, args),
-                      partial(Network.SIPpNetwork.shutdown, self.network)]:
+            try:
                 try:
-                    h()
-                except BaseException as e:
-                    self.__get_logger().debug(e, exc_info = True)
-                    state = SIPpTest.State.DIRTY
-
-            self.__set_state(state)
-            if state == SIPpTest.State.DIRTY:
+                    self.__run_script("after.sh", args)
+                finally:
+                    self.network.sniffer_stop()
+                    self.__remove_temp_folder(args)
+                    self.network.shutdown()
```

If `self.network.sniffer_stop()` raises an exception, `self.__remove_temp_folder(args)` and  `self.network.shutdown()` will not be executed. And I'm not sure, but I think the first exception will be lost and the catched exception outside (line 405) will be the last one.

Is this ok?

---

Great catch, Danny! I re-worked the implementation